Dockerfile

```diff
@@ -0,0 +1,156 @@
+###################################################################################################
```

Comment:
This will be a useful addition for reproducibility, also I'm completely fine with pinning the dependencies in `requirements.txt` since we already have the loose versions in `setup.py`

Reply:
I view the value of pinning dependencies as dependent on maintenance plans for the repo.  It's often useful to have package versions pinned if they'll be updated often, and it can be significantly misleading in some cases if there's no plan to maintain them (though still useful as a record of when the last work was done).   I was trying to take a low impact approach here, and happy for others to iterate on it as desired.

Also if Docker images eventually get pushed to DockerHub, they will include dependency versions that were in use at the time, so provide another layer of (implicit-but-recoverable) versioning.

Reply:
I'm fine either was but in my experience you usually want pinned versions in a Dockerfile and loose versions in a library that will be pip installed. That way the library is flexible and can be incorporated into larger projects but the dockerfile is stable and reproducible.

With that said, I know I have held this up for a long time we can come back to this down the road if we run into dependency problems.

Reply:
That sounds like a good approach to revisit.   I think there's a longer conversation in some cases -- e.g. pinning the Jupyter version will likely trigger some dependabot security alerts if it goes out of date over time.